client/src/client.rs

```diff
@@ -313,43 +480,98 @@ impl Client {
             )
     }
 
-    /// Query API entry point. Requests queries from `Iroha` peers with pagination.
+    /// Lower-level Query API entry point. Prepares an http-request and returns it with an http-response handler.
     ///
     /// # Errors
-    /// Fails if sending request fails
-    #[log]
-    pub fn request_with_pagination<R>(
+    /// Fails if query signing fails.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// use eyre::Result;
+    /// use iroha_client::{
+    ///     client::{Client, ResponseHandler},
+    ///     http::{RequestBuilder, Response},
+    /// };
+    /// use iroha_data_model::prelude::{Account, FindAllAccounts, Pagination};
+    ///
+    /// struct YourAsyncRequest;
+    ///
+    /// impl YourAsyncRequest {
+    ///     async fn send(self) -> Response<Vec<u8>> {
+    ///         // do the stuff
+    ///     }
+    /// }
+    ///
+    /// // Implement builder for this request
+    /// impl RequestBuilder for YourAsyncRequest {
+    ///     // ...
+    /// }
+    ///
+    /// async fn fetch_accounts(client: &Client) -> Result<Vec<Account>> {
+    ///     // Put `YourAsyncRequest` as a type here
+    ///     // It returns the request and the handler (zero-cost abstraction) for the response
+    ///     let (req, resp_handler) = client.prepare_query_request::<_, YourAsyncRequest>(
+    ///         FindAllAccounts::new(),
+    ///         Pagination::default(),
+    ///     )?;
+    ///
+    ///     // Do what you need to send the request and to get the response
+    ///     let resp = req.send().await;
+    ///
+    ///     // Handle response with the handler and get typed result
+    ///     let accounts = resp_handler.handle(resp)?;
+    ///
+    ///     Ok(accounts.only_output())
+    /// }
+    /// ```
+    pub fn prepare_query_request<R, B>(
         &self,
         request: R,
         pagination: Pagination,
-    ) -> Result<R::Output>
+    ) -> Result<(B, QueryResponseHandler<R>)>
```

Comment:
@s8sato, intermediate structs are for convenience. Especially with `QueryResponseHandler<R>` - you could not care about the correct output typing, it is just seamlessly incapsulated within the handler, and you need just to call `.handle(response)` to get the output. Other handlers aren't typed, but they still help. You don't have to import a particular handling function, you just need to pass the response to the handler. It makes flow a bit more clear.

#1256 is resolved by this PR **only partially**. I don't think that putting the concrete HTTP-client implementation like `hyper::client` or `reqwest` is a good idea. If to be concrete - on iroha2-block-explorer I use `actix-web`, and Actix ecosystem delivers its own HTTP client. It would be not good if iroha-client uses some other http-client instead of it.

My solution is more flexible, but in cost of complexity. You can use it with any HTTP-transport you want, async or not, but you have to implement integration of your transport into iroha client.

Btw, it is possible to create out-of-the-box implementations for popular clients **on top** of my solution, and then export them from the client library (e.g. using feature flags).

Reply:
Okay, `actix-web` seems to use `awc` client.
I think I can see how `RequestBuilder` and `ResponseHandler`s will be utilized in subsequent PRs